docs/designs/odata-writer-rewrite-proposal.md

```diff
@@ -0,0 +1,315 @@
+
+- What layers of abstractions should the new writer have to ensure we offer sufficient flexibility for use cases the OData writer must support, while providing efficient default implementations for common scenarios?
+- How tightly coupled should we be with `JsonSerializer` and `Utf8JsonWriter`?
+- Should we account for other formats other than JSON?
+- Should we require an `IEdmModel` to be provided for serialization?
+
+Here are the design principles that I want to adhere to:
+
+- Avoid forcing the caller to create intermediate data unnecessarily.
+  - A lot of the cost of the existing writer comes from intermediate data and repeated/redundant computations.
+  - The existing writer requires the caller to create intermediate data via `ODataResource`, `ODataValue`, etc.
+- If the caller passes structs as input, they should not be forced to box them.
+  - Or existing abstractions often accept or return `object`, which force value types to be boxed.
+- Avoid paying the cost for things we don't use.
+  - Existing writer allocates things that are not used (e.g. `ODataResource.Actions`).
+  - Existing writer performs validation eagerly (e.g. `ODataResourceBase.VerifyProperties()`) even when not necessary.
+- New writer should expose similar levels of abstractions and extensibility to existing writer
+- New writer should provide support for all the scenarios that the existing writer supports
+  - Different input types (CLR objects, EdmObject, Dictionaries, JObject)
+  - Different types of payloads: resource set, values, aggregation results, service document, raw values, metadata document, etc.
+  - Dynamic properties, delta payloads, stream properties, polymorphic properties, $select/$expand
+  - Standard OData annotations, custom annotations, etc.
+  - Streaming support (start serialization before all data is available)
+
+The sample code in this doc or for demonstration purposes and are subject to change as the design is fleshed out more.
+
+## Top-level writer
+
+This is the equivalent to the existing `ODataMessageWriter` layer, which also exposes the abstract `ODataWriter` and `ODataOutputContext` classes.
+The `ODataMessageWriter` exposes methods for writing top-level payloads like metadata, service document, property values, etc. or methods
+for creating writers for writing more complex, nested payloads like resources and resouce sets.
+
+However, `ODataMessageWriter` doesn't implement any logic, all the logic is delegated to the abstract `ODataOutputContext` and `ODataWriter` classes.
+The `ODataOutputContext` does most of the heavy lifting and is the one that determines the output format.
+
+This means that if you wanted to create your own customer `ODataMessageWriter` from scratch (e.g. to output a different format other than JSON), you would
+need to provide custom `ODataOutputContext` and `ODataWriter` implementations. This is a non-trivial undertaking and I doubt there are any implementations

```

Comment:
Check this: https://devblogs.microsoft.com/odata/customize-odata-payload-serialization-format-within-asp-net-core-odata/